File: Anime Game Remap (for all users)/api/src/py/FixRaidenBoss2/model/IniNamingTools.py
##### Credits

# ===== Anime Game Remap (AG Remap) =====
# Authors: Albert Gold#2696, NK#1321
#
# if you used it to remap your mods pls give credit for "Albert Gold#2696" and "Nhok0169"
# Special Thanks:
#   nguen#2011 (for support)
#   SilentNightSound#7430 (for internal knowdege so wrote the blendCorrection code)
#   HazrateGolabi#1364 (for being awesome, and improving the code)

##### EndCredits

##### ExtImports
import os
import re
import pathlib
from typing import Tuple, Optional
##### EndExtImports

##### LocalImports
from ..constants.FileExt import FileExt
from ..constants.IniConsts import IniKeywords
from ..tools.TextTools import TextTools
##### EndLocalImports


##### Script
class IniNamingTools():
    """
    Utilities for some common naming conventions for .ini files
    """

    @classmethod
    def getResourceName(cls, name: str) -> str:
        """
        Makes the name of a `section`_ to be used for the resource `sections`_ of a .ini file

        Examples
        --------
        >>> IniNamingTools.getResourceName("CuteLittleEi")
        "ResourceCuteLittleEi"


        >>> IniNamingTools.getResourceName("ResourceCuteLittleEi")
        "ResourceCuteLittleEi"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        Returns
        -------
        :class:`str`
            The name of the `section`_ as a resource in a .ini file
        """

        if (not name.startswith(IniKeywords.Resource.value)):
            name = f"{IniKeywords.Resource.value}{name}"
        return name
    
    @classmethod
    def removeResourceName(cls, name: str) -> str:
        """
        Removes the 'Resource' prefix from a section's name

        Examples
        --------
        >>> IniNamingTools.removeResourceName("ResourceCuteLittleEi")
        "CuteLittleEi"


        >>> IniNamingTools.removeResourceName("LittleMissGanyu")
        "LittleMissGanyu"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the 'Resource' prefix removed
        """

        if (name.startswith(IniKeywords.Resource.value)):
            name = name[len(IniKeywords.Resource.value):]

        return name
    
    @classmethod
    def getRemapElementName(cls, name: str, elementName: str, modName: str = ""):
        """
        Changes a `section`_ name to have the keyword from 'elementName' to identify that the `section`_
        is created by this fix

        Examples
        --------
        >>> IniNamingTools.getRemapElementName("EiTriesToUseBlenderAndFails", "Blend", "Raiden")
        "EiTriesToUseRaidenRemapBlenderAndFails"


        >>> IniNamingTools.getRemapElementName("EiTextsTheTexture", "Tex", "Yae")
        "EiTextsTheYaeRemapTexture"
    

        >>> IniNamingTools.getRemapElementName("ResourceCuteLittleEi", "Position", "Raiden")
        "ResourceCuteLittleEiRaidenRemapPosition"


        >>> IniNamingTools.getRemapElementName("ResourceCuteLittleEiRemapDango", "Dango" "Raiden")
        "ResourceCuteLittleEiRemapRaidenRemapDango"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        elementName: :class:`str`
            The name of the target element

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the keyword of 'elementName', prefixed by the word 'Remap' added
        """

        nameParts = name.rsplit(elementName, 1)
        namePartsLen = len(nameParts)

        remapName = f"{modName}{IniKeywords.Remap.value}{elementName}"
        if (namePartsLen > 1):
            name = remapName.join(nameParts)
        else:
            name += remapName

        return name
    
    @classmethod
    def getRemapBlendName(cls, name: str, modName: str = "") -> str:
        """
        Changes a `section`_ name to have the keyword 'RemapBlend' to identify that the `section`_
        is created by this fix

        .. tip::
            See :meth:`getRemapElementName` for some examples

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapBlend' keyword
        """

        return cls.getRemapElementName(name, elementName = IniKeywords.Blend.value, modName = modName)
    
    @classmethod
    def getRemapPositionName(cls, name: str, modName: str = "") -> str:
        """
        Changes a `section`_ name to have the keyword 'RemapPosition' to identify that the `section`_
        is created by this fix

        .. tip::
            See :meth:`getRemapElementName` for some examples

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapPosition' keyword
        """

        return cls.getRemapElementName(name, elementName = IniKeywords.Position.value, modName = modName)
    
    @classmethod
    def getRemapTexcoordName(cls, name: str, modName: str = "") -> str:
        """
        Changes a `section`_ name to have the keyword 'RemapTexcoord' to identify that the `section`_
        is created by this fix

        .. tip::
            See :meth:`getRemapElementName` for some examples

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapTexcoord' keyword
        """

        return cls.getRemapElementName(name, elementName = IniKeywords.Texcoord.value, modName = modName)
    
    @classmethod
    def getRemapIbName(cls, name: str, modName: str = "") -> str:
        """
        Changes a `section`_ name to have the keyword 'RemapIb' to identify that the `section`_
        is created by this fix

        .. tip::
            See :meth:`getRemapElementName` for some examples

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapIb' keyword
        """

        return cls.getRemapElementName(name, elementName = "IB", modName = modName)
    
    @classmethod
    def getModSuffixedName(cls, name: str, suffix: str = "", modName: str = ""):
        """
        Changes a `section`_ name to have the suffix of 'modName' followed by 'suffix'

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        suffix: :class:`str`
            The name of the suffix to put at the end of the `section`_ :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added suffix keyword
        """

        remapName = f"{modName}{suffix}"
        if (name.endswith(remapName)):
            return name
        elif(name.endswith(suffix)):
            return name[:len(suffix)] + remapName

        return name + remapName
    
    @classmethod
    def getRemapFixName(cls, name: str, modName: str = "") -> str:
        """
        Changes a `section`_ name to have the suffix `RemapFix` to identify that the `section`_
        is created by this fix

        Examples
        --------
        >>> IniNamingTools.getRemapFixName("EiIsDoneWithRemapFix", "Raiden")
        "EiIsDoneWithRaidenRemapFix"

        >>> IniNamingTools.getRemapFixName("EiIsHappy", "Raiden")
        "EiIsHappyRaidenRemapFix"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapFix' keyword
        """

        return cls.getModSuffixedName(name, suffix = IniKeywords.RemapFix.value, modName = modName)
    
    @classmethod
    def getRemapTexName(cls, name: str, modName: str = ""):
        """
        Changes a `section`_ name to have the keyword 'RemapTex' to identify that the `section`_
        is created by this fix

        Examples
        --------
        >>> IniNamingTools.getRemapTexName("EiIsDoneWithRemapTex", "Raiden")
        "EiIsDoneWithRaidenRemapTex"

        >>> IniNamingTools.getRemapTexName("EiIsHappy", "Raiden")
        "EiIsHappyRaidenRemapTex"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapFix' keyword
        """

        return cls.getModSuffixedName(name, suffix = IniKeywords.RemapTex.value, modName = modName)
    
    @classmethod
    def getRemapDLName(cls, name: str, modName: str = ""):
        """
        Changes a `section`_ name to have the suffix `RemapDL` to identify that the `section`_
        is created by this fix

        Examples
        --------
        >>> IniNamingTools.getRemapTexName("EiIsDoneWithRemapDL", "Raiden")
        "EiIsDoneWithRaidenRemapDL"

        >>> IniNamingTools.getRemapTexName("EiIsHappy", "Raiden")
        "EiIsHappyRaidenRemapDL"

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the `section`_ with the added 'RemapDL' keyword
        """

        return cls.getModSuffixedName(name, suffix = IniKeywords.RemapDL.value, modName = modName)

    @classmethod
    def getRemapFixResourceName(cls, name: str, modName: str = ""):
        """
        Changes a `section`_ name to be a new non-blend resource created by this fix

        .. note::
            See :meth:`getResourceName` and :meth:`getRemapFix` for more info

        Parameters
        ----------
        name: :class:`str`
            The name of the section

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the section with the prefix 'Resource' and the suffix 'RemapFix' added
        """

        name = cls.getRemapFixName(name, modName = modName)
        name = cls.getResourceName(name)
        return name
    
    @classmethod
    def getRemapTexResourceName(cls, name: str, modName: str = ""):
        """
        Changes a `section`_ name to be a texture resource created by this fix

        .. note::
            See :meth:`getResourceName` and :meth:`getRemapTexName` for more info

        Parameters
        ----------
        name: :class:`str`
            The name of the section

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the section with the prefix 'Resource' and the suffix 'RemapTex' added
        """

        name = cls.getRemapTexName(name, modName = modName)
        name = cls.getResourceName(name)
        return name
    
    @classmethod
    def getRemapDLResourceName(cls, name: str, modName: str = ""):
        """
        Changes a `section`_ name to be a texture resource created by this fix

        .. note::
            See :meth:`getResourceName` and :meth:`getRemapDLName` for more info

        Parameters
        ----------
        name: :class:`str`
            The name of the section

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the section with the prefix 'Resource' and the suffix 'RemapDL' added
        """

        name = cls.getRemapDLName(name, modName = modName)
        name = cls.getResourceName(name)
        return name

    @classmethod
    def getRemapBlendResourceName(cls, name: str, modName: str = "") -> str:
        """
        Changes the name of a section to be a new blend resource that this fix will create

        .. note::
            See :meth:`getResourceName` and :meth:`getRemapBlendName` for more info

        Parameters
        ----------
        name: :class:`str`
            The name of the section

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the section with the prefix 'Resource' and the keyword 'Remap' added
        """

        name = cls.getRemapBlendName(name, modName = modName)
        name = cls.getResourceName(name)
        return name
    
    @classmethod
    def getRemapPositionResourceName(cls, name: str, modName: str = "") -> str:
        """
        Changes the name of a section to be a new position resource that this fix will create

        .. note::
            See :meth:`getResourceName` and :meth:`getRemapPositionName` for more info

        Parameters
        ----------
        name: :class:`str`
            The name of the section

        modName: :class:`str`
            The name of the mod to fix :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name of the section with the prefix 'Resource' and the keyword 'Remap' added
        """

        name = cls.getRemapPositionName(name, modName = modName)
        name = cls.getResourceName(name)
        return name
    
    @classmethod
    def getFixedFile(cls, file: str, modName: str = "", fileExt: Optional[str] = None) -> str:
        """
        Retrieves the file path for a a fixed element

        Parameters
        ----------
        file: :class:`str`
            The file path to the original file

        modName: :class:`str`
            The name of the mod to fix to :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        fileExt: Optional[:class:`str`]
            The file extension for the file path of the fixed element.
            If this argument is ``None``, then will use the file extension provided in the file :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns
        -------
        :class:`str`
            The file path of the fixed file of the element
        """

        path = pathlib.Path(file)
        folder = path.parent
        baseName = path.stem
        
        if (fileExt is None):
            fileExt = path.suffix
        
        return os.path.join(f"{folder}", f"{cls.getRemapFixName(baseName, modName = modName)}{fileExt}")
    
    @classmethod
    def getFixedElementFile(cls, file: str, elementName: str, modName: str = "", fileExt: Optional[str] = None) -> str:
        """
        Retrieves the file path for a a fixed element

        Parameters
        ----------
        file: :class:`str`
            The file path to the original file

        elementName: :class:`str`
            The name of the element to fix

        modName: :class:`str`
            The name of the mod to fix to :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        fileExt: Optional[:class:`str`]
            The file extension for the file path of the fixed element.
            If this argument is ``None``, then will use the file extension provided in the file :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``None``

        Returns
        -------
        :class:`str`
            The file path of the fixed file of the element
        """

        path = pathlib.Path(file)
        folder = path.parent
        baseName = path.stem
        
        if (fileExt is None):
            fileExt = path.suffix

        file = f"{cls.getRemapElementName(baseName, elementName, modName = modName)}{fileExt}"
        if (folder == pathlib.Path(".")):
            return file
        
        return os.path.join(f"{folder}", file)

    @classmethod
    def getFixedBlendFile(cls, blendFile: str, modName: str = "") -> str:
        """
        Retrieves the file path for the fixed RemapBlend.buf file

        Parameters
        ----------
        blendFile: :class:`str`
            The file path to the original Blend.buf file

        modName: :class:`str`
            The name of the mod to fix to :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The file path of the fixed RemapBlend.buf file
        """

        return cls.getFixedElementFile(blendFile, IniKeywords.Blend.value, modName = modName, fileExt = FileExt.Buf.value)
    
    @classmethod
    def getFixedPositionFile(cls, positionFile: str, modName: str = "") -> str:
        """
        Retrieves the file path for the fixed RemapPosition.buf file

        Parameters
        ----------
        positionFile: :class:`str`
            The file path to the original Position.buf file

        modName: :class:`str`
            The name of the mod to fix to

        Returns
        -------
        :class:`str`
            The file path of the fixed RemapPosition.buf file
        """

        return cls.getFixedElementFile(positionFile, IniKeywords.Position.value, modName = modName, fileExt = FileExt.Buf.value)
    
    @classmethod
    def getFixedTexFile(cls, texFile: str, modName: str = "") -> str:
        """
        Retrieves the file path for the fixed RemapTex.dds file

        Parameters
        ----------
        texFile: :class:`str`
            The file path to the original .dds file

        modName: :class:`str`
            The name of the mod to fix to :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The file path of the fixed RemapTex.dds file
        """

        blendFolder = os.path.dirname(texFile)
        blendBaseName = os.path.basename(texFile)
        blendBaseName = blendBaseName.rsplit(".", 1)[0]

        return os.path.join(blendFolder, f"{cls.getRemapTexName(blendBaseName, modName = modName)}{FileExt.DDS.value}")
    
    @classmethod
    def getTextureOverrideRemapFix(cls, component: str, obj: str, modName: str = ""):
        """
        Retrieves the name to some generic ``TextureOverride`` `section`_ this software has made

        Parameters
        ----------
        component: :class:`str`
            The name of the component

        obj: :class:`str`
            The name of the object

        modName: :class:`str`
            The name of the mod  :raw-html:`<br />` :raw-html:`<br />`

            **Default**: ``""``

        Returns
        -------
        :class:`str`
            The name for the `section`_
        """

        return cls.getRemapFixName(f"{IniKeywords.TextureOverride.value}{TextTools.capitalize(modName)}{TextTools.capitalize(component)}{TextTools.capitalize(obj)}")
    
    @classmethod
    def getObjRemapFixName(cls, name: str, modName: str, objName: Tuple[str, str], newObjName: Tuple[str, str]) -> str:
        """
        Retrieves the new name of the `section`_ for a new mod object

        Parameters
        ----------
        name: :class:`str`
            The name of the `section`_

        modName: :class:`str`
            The name of the mod to be fixed

        objName: Tuple[:class:`str`, :class:`str`]
            The name of the component and object for the original mod object for the `section`_

        newObjName: Tuple[:class:`str`, :class:`str`]
            The name of the component and object for the new mod object for the `section`_

        Returns
        -------
        :class:`str`
            The new name for the `section`_
        """


        name = TextTools.reverse(name)
        objName = f"{TextTools.capitalize(objName[0])}{TextTools.capitalize(objName[1])}"
        newObjName = f"{TextTools.capitalize(newObjName[0])}{TextTools.capitalize(newObjName[1])}"
        modName = TextTools.capitalize(modName)
    
        nameParts = re.split(TextTools.reverse(objName), name, flags = re.IGNORECASE, maxsplit = 1)
        namePartsLen = len(nameParts)

        if (namePartsLen == 1):
            name = TextTools.reverse(name)
            return cls.getRemapFixName(name, modName = f"{modName}{newObjName}")

        name = TextTools.reverse(newObjName).join(nameParts)
        name = TextTools.reverse(name)
        return cls.getRemapFixName(name, modName = modName)
##### EndScript

File: Anime Game Remap (for all users)/api/src/cpp/core/src/model/IniNamingTools.cpp
#include "AGRemapCore/model/IniNamingTools.h"

#include <cctype>
#include <filesystem>
#include <optional>

#include "AGRemapCore/constants/FileExt.h"
#include "AGRemapCore/constants/IniKeywords.h"
#include "AGRemapCore/tools/TextTools.h"


namespace AGRemapCore {

    namespace {
        namespace fs = std::filesystem;

        bool asciiIEquals(char a, char b) {
            return std::tolower(static_cast<unsigned char>(a)) == std::tolower(static_cast<unsigned char>(b));
        }

        // Case-insensitive (ASCII) search for the LAST occurrence of 'needle' in 'haystack' --
        // mirrors what the pure-Python original's getObjRemapFixName achieves indirectly (reverse
        // both strings, then re.split(..., maxsplit = 1) from the front is equivalent to finding
        // the last match from the end of the un-reversed string; verified empirically against the
        // Python original before simplifying it down to this direct form). ASCII-only, matching
        // this codebase's existing precedent for fixed-shape identifier matching
        // (IfPredPartTypeTools's toLowerAscii) rather than pulling in full Unicode case-folding for
        // what are, in practice, always ASCII component/mod names.
        std::optional<size_t> rfindCaseInsensitiveAscii(std::string_view haystack, std::string_view needle) {
            if (needle.empty() || needle.size() > haystack.size()) {
                return std::nullopt;
            }

            for (size_t i = haystack.size() - needle.size() + 1; i-- > 0; ) {
                bool match = true;
                for (size_t j = 0; j < needle.size(); ++j) {
                    if (!asciiIEquals(haystack[i + j], needle[j])) {
                        match = false;
                        break;
                    }
                }
                if (match) {
                    return i;
                }
            }

            return std::nullopt;
        }

        // pathlib.Path(file).parent always returns Path(".") (never an empty path) for a bare
        // filename with no directory component -- unlike std::filesystem::path::parent_path(),
        // which returns an empty path for the same input. getFixedFile/getFixedElementFile below
        // rely on this "." fallback (the pure-Python original's own getFixedElementFile even checks
        // "folder == pathlib.Path('.')" explicitly). getFixedTexFile, built on
        // os.path.dirname/os.path.join instead, deliberately does NOT get this treatment --
        // os.path.dirname returns "" for a bare filename, and os.path.join("", x) == x with no
        // "./" prefix at all -- see its own comment.
        fs::path pathlibStyleParent(const fs::path& path) {
            fs::path parent = path.parent_path();
            return parent.empty() ? fs::path(".") : parent;
        }
    }

    std::string IniNamingTools::getResourceName(const std::string& name) {
        if (!name.starts_with(IniKeywords::Resource)) {
            return IniKeywords::Resource + name;
        }
        return name;
    }

    std::string IniNamingTools::removeResourceName(const std::string& name) {
        if (name.starts_with(IniKeywords::Resource)) {
            return name.substr(IniKeywords::Resource.size());
        }
        return name;
    }

    std::string IniNamingTools::getRemapElementName(const std::string& name, const std::string& elementName, const std::string& modName) {
        std::string remapName = modName + IniKeywords::Remap + elementName;

        if (elementName.empty()) {
            return name + remapName;
        }

        // Replaces the LAST occurrence of 'elementName' with 'remapName' -- equivalent to the
        // Python original's "name.rsplit(elementName, 1)" (split on the last occurrence) followed
        // by "remapName.join(nameParts)".
        size_t pos = name.rfind(elementName);
        if (pos == std::string::npos) {
            return name + remapName;
        }

        return name.substr(0, pos) + remapName + name.substr(pos + elementName.size());
    }

    std::string IniNamingTools::getRemapBlendName(const std::string& name, const std::string& modName) {
        return getRemapElementName(name, IniKeywords::Blend, modName);
    }

    std::string IniNamingTools::getRemapPositionName(const std::string& name, const std::string& modName) {
        return getRemapElementName(name, IniKeywords::Position, modName);
    }

    std::string IniNamingTools::getRemapTexcoordName(const std::string& name, const std::string& modName) {
        return getRemapElementName(name, IniKeywords::Texcoord, modName);
    }

    std::string IniNamingTools::getRemapIbName(const std::string& name, const std::string& modName) {
        // Literal "IB", matching the Python original exactly -- NOT IniKeywords::Ib (which would be
        // lowercase "ib"; that member isn't ported here since nothing else in this class uses it).
        return getRemapElementName(name, "IB", modName);
    }

    std::string IniNamingTools::getModSuffixedName(const std::string& name, const std::string& suffix, const std::string& modName) {
        std::string remapName = modName + suffix;

        if (name.ends_with(remapName)) {
            return name;
        }

        if (name.ends_with(suffix)) {
            // The pure-Python original has a confirmed bug here: it returns
            // "name[:len(suffix)] + remapName" (the first len(suffix) characters of 'name') instead
            // of "name[:-len(suffix)] + remapName" (name with the trailing 'suffix' stripped off) --
            // e.g. getRemapFixName("EiIsDoneWithRemapFix", "Raiden") returns
            // "EiIsDoneRaidenRemapFix" in Python, contradicting that very method's own docstring
            // example of "EiIsDoneWithRaidenRemapFix". Confirmed by running the Python original
            // directly. Per the maintainer (asked explicitly), this C++ port implements the
            // documented/intended behavior below, not the buggy one.
            return name.substr(0, name.size() - suffix.size()) + remapName;
        }

        return name + remapName;
    }

    std::string IniNamingTools::getRemapFixName(const std::string& name, const std::string& modName) {
        return getModSuffixedName(name, IniKeywords::RemapFix, modName);
    }

    std::string IniNamingTools::getRemapTexName(const std::string& name, const std::string& modName) {
        return getModSuffixedName(name, IniKeywords::RemapTex, modName);
    }

    std::string IniNamingTools::getRemapDLName(const std::string& name, const std::string& modName) {
        return getModSuffixedName(name, IniKeywords::RemapDL, modName);
    }

    std::string IniNamingTools::getRemapFixResourceName(const std::string& name, const std::string& modName) {
        return getResourceName(getRemapFixName(name, modName));
    }

    std::string IniNamingTools::getRemapTexResourceName(const std::string& name, const std::string& modName) {
        return getResourceName(getRemapTexName(name, modName));
    }

    std::string IniNamingTools::getRemapDLResourceName(const std::string& name, const std::string& modName) {
        return getResourceName(getRemapDLName(name, modName));
    }

    std::string IniNamingTools::getRemapBlendResourceName(const std::string& name, const std::string& modName) {
        return getResourceName(getRemapBlendName(name, modName));
    }

    std::string IniNamingTools::getRemapPositionResourceName(const std::string& name, const std::string& modName) {
        return getResourceName(getRemapPositionName(name, modName));
    }

    std::string IniNamingTools::getFixedFile(const std::string& file, const std::string& modName, std::optional<std::string> fileExt) {
        fs::path path(file);
        fs::path folder = pathlibStyleParent(path);
        std::string baseName = path.stem().string();
        std::string ext = fileExt.has_value() ? *fileExt : path.extension().string();

        std::string newName = getRemapFixName(baseName, modName) + ext;
        return (folder / newName).string();
    }

    std::string IniNamingTools::getFixedElementFile(const std::string& file, const std::string& elementName, const std::string& modName, std::optional<std::string> fileExt) {
        fs::path path(file);
        fs::path folder = pathlibStyleParent(path);
        std::string baseName = path.stem().string();
        std::string ext = fileExt.has_value() ? *fileExt : path.extension().string();

        std::string newName = getRemapElementName(baseName, elementName, modName) + ext;
        if (folder == fs::path(".")) {
            return newName;
        }

        return (folder / newName).string();
    }

    std::string IniNamingTools::getFixedBlendFile(const std::string& blendFile, const std::string& modName) {
        return getFixedElementFile(blendFile, IniKeywords::Blend, modName, FileExt::Buf);
    }

    std::string IniNamingTools::getFixedPositionFile(const std::string& positionFile, const std::string& modName) {
        return getFixedElementFile(positionFile, IniKeywords::Position, modName, FileExt::Buf);
    }

    std::string IniNamingTools::getFixedTexFile(const std::string& texFile, const std::string& modName) {
        fs::path path(texFile);
        fs::path folder = path.parent_path();  // no "." fallback here -- see pathlibStyleParent's comment
        std::string baseName = path.filename().string();

        // Matches Python's "basename.rsplit('.', 1)[0]" -- strip only the LAST "." extension,
        // keeping everything before it (including any earlier dots). NOT the same as
        // std::filesystem::path::stem(), which parses extensions with filesystem-style rules (eg. a
        // leading-dot-only "dotfile" like ".gitignore" has no extension to std::filesystem, but
        // rsplit('.', 1) still splits it into "" + "gitignore"). Kept as a manual rsplit to match
        // the Python original exactly rather than std::filesystem::path::stem()'s edge-case rules.
        size_t dotPos = baseName.rfind('.');
        if (dotPos != std::string::npos) {
            baseName = baseName.substr(0, dotPos);
        }

        std::string newName = getRemapTexName(baseName, modName) + FileExt::DDS;
        return (folder / newName).string();
    }

    std::string IniNamingTools::getTextureOverrideRemapFix(const std::string& component, const std::string& obj, const std::string& modName) {
        return getRemapFixName(IniKeywords::TextureOverride + TextTools::capitalize(modName) + TextTools::capitalize(component) + TextTools::capitalize(obj));
    }

    std::string IniNamingTools::getObjRemapFixName(const std::string& name, const std::string& modName,
                                                    const std::pair<std::string, std::string>& objName,
                                                    const std::pair<std::string, std::string>& newObjName) {
        std::string objNameStr = TextTools::capitalize(objName.first) + TextTools::capitalize(objName.second);
        std::string newObjNameStr = TextTools::capitalize(newObjName.first) + TextTools::capitalize(newObjName.second);
        std::string capModName = TextTools::capitalize(modName);

        std::optional<size_t> pos = rfindCaseInsensitiveAscii(name, objNameStr);
        if (!pos.has_value()) {
            return getRemapFixName(name, capModName + newObjNameStr);
        }

        std::string newName = name.substr(0, *pos) + newObjNameStr + name.substr(*pos + objNameStr.size());
        return getRemapFixName(newName, capModName);
    }
}
